qbplatform/quote/models.py

```diff
@@ -13,19 +15,59 @@ class Tournament(models.Model):
 
     name = models.CharField(max_length=500)
 
+    def __str__(self):
+        return self.name
+
 
 class TournamentLocation(models.Model):
 
     tournament = models.ForeignKey(Tournament, on_delete=models.CASCADE)
     date = models.DateField()
     location = models.CharField(max_length=500, null=True)
 
+    def __str__(self):
+        return self.location + " site of tournament: " + str(self.tournament)
+
 
 class Packet(models.Model):
 
     name = models.CharField(max_length=500)
     tournament = models.ForeignKey(Tournament, on_delete=models.CASCADE)
 
+    def __str__(self):
+        return "Packet " + self.name + " in tournament: " + str(self.tournament)
+
+#   Model of a per-packet distribution
+#       has a one-to-many relationship with DistributionCategory
+class Distribution(models.Model):
+
+    num_tossups = models.PositiveSmallIntegerField(null=True, default=20)
+    num_bonuses = models.PositiveSmallIntegerField(null=True, default=20)
+
+    # a distribution is unique to a tournament/question set 
+    tournament = models.OneToOneField(Tournament, null=True, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return str(self.tournament) + " distribution"
+
+#   Model of a category in a per-packet distribution
+#       is a hierarchy from top level category (e.g. Literature) to subcategory (e.g. American Literature) 
+#
+#       each (sub)category is allowed a specified range of tossups/bonuses per packet
+#           for example, an editor could specify 1/1 Other Arts but 0-1/0-1 Architecture
+class DistributionCategory(TreeNode):
```

Comment:
One concern that I had about this is that this structure tightly couples the category logic to the distribution. What I mean by that is that, as constructed, this model requires that if I have a Distribution A and a Distribution B and they both have e.g. Literature as an identical branch of the category tree but differ in Science, that means that they have to be independently defined. This means that there is no such thing in our ontology as "Literature" per se; there is only "Literature, as specified in the ACF Distribution," "Literature, as specified in My Fancy Lit Tournament Distribution," etc. Maybe this isn't a problem! I don't anticipate that there will be hundreds of distributions, and maybe it doesn't matter if there's no "coherence" among them. But I think I would like to see if we can avoid this very tight coupling because it feels rigid and also a bit denormalizing. What if we had something like a hierarchical category structure, implemented with `TreeNode`, that carried the actual name, and then `DistributionCategory` had a `ManyToMany` relationship with the category itself? Then categories could be defined independently of distributions (and have a hierarchical relationship amongst themselves) but we could attach to every `DistributionCategory` a `Category`  that contained the actual semantics of what we want. Now we have a separation between the categories and the requirements of the categories in any given distribution, and we can compare things in categories across tournaments too, if we want.

Reply:
Should be addressed in [ebfc59f](https://github.com/grapesmoker/qb-question-platform/pull/23/commits/ebfc59f352ab3300e977fed9aa28cbe2050af95d).

One thing, though, with this current setup, is that, since DistributionCategory is no longer hierarchical, a DistributionCategory has no relationship to any other DistributionCategory. I'm not sure that's ideal? Is it worth having DistributionCategory extend TreeNode as well?

Reply:
Right, I think I was trying to make this point when we talked on Sunday but I started wandering and lost the plot. If we made `DistributionCategory` a hierarchical structure, then we could have have subdistributional requirements without changing anything about the structure. Otherwise, if we wanted to require e.g. 4/4 science and also 1/1 physics we would have to have some kind of flat scheme for attaching that to a distribution. So I think the correct move is to make this a `TreeNode` and allow every level to specify numerical min/max requirements.